### PR TITLE
Frontend segmentation components

### PR DESCRIPTION
**Pull request by DamienGilliard**
This PR concerns the segmentation of the scan. We use the 3d model to identify the pieces we need to segment. It introduces both the cpp backend and components for:

- [x] basic normal based segmentation
- [x] CAD-based segmentation

### CODE REVIEW

**Comment by DamienGilliard** [comment]
Proposed strategy for additive:

Capture d’écran 2024-06-01 à 14 16 18

**Comment by 9and3** [comment]
thanks for the  proposal @DamienGilliard ! 🚀 

my questions/doubts:
- are you sure about the recursive registration to the scan, might be time consuming but especially not robust (?)
- what about stop to global+refine and segmenting the cloud by normals (e.g. via cilantro) and than do a box search with the elements of the CAD model to find if the center of the clustes lies inside (or something similar) ?

**Comment by DamienGilliard** [comment]
Thanks for the reply ! 

I share your doubts about the recursive registration, i will soon know if it works sufficiently well (computation time wise, and accuracy wise)

For the second point, the reason is that planar surfaces in the scan can be the result of different pieces combined. This means that plane segmentation is not sufficient to get the point clouds of each piece. Do you agree ? I might be missing something ;)

**Comment by 9and3** [comment]
> For the second point, the reason is that planar surfaces in the scan can be the result of different pieces combined. This means that plane segmentation is not sufficient to get the point clouds of each piece. Do you agree

Yes you are right indeed. Got no solutions for that though..

**Comment by DamienGilliard** [comment]
This branch has now been cleaned and contains the implementation of the initial segmentation. Branch(es) of this branch will contain the further implementation of the "semantic" segmentation.
Capture d'écran 2024-06-23 124431

File: src/diffCheck/segmentation/DFSegmentation.cc
Comment (9and3) [inline src/diffCheck/segmentation/DFSegmentation.cc:1]: NormalBasedSegmentation would not be more appropiate? Why "Smooth"? @DamienGilliard

Comment (DamienGilliard) [reply src/diffCheck/segmentation/DFSegmentation.cc:1]: done !

Comment (DamienGilliard) [reply src/diffCheck/segmentation/DFSegmentation.cc:1]: the rebase of the branch of this branch went so smoothly I doubted it had worked 👌

**Comment by DamienGilliard** [comment]
image
on rougher point clouds (AKA real world point clouds), the plane segmentation leads to unsatisfactory results and will require additional steps to reach good results. Re-introducing an association distance threshold so we can only associate the closest points of the segments, instead of all points.

**Comment by DamienGilliard** [comment]
@9and3  add:
* conversion of point cloud from DFPointCloud to cilantroPointCloud
* separate downsampling and normal estimation outside of cilantro plane segmentation

**Comment by 9and3** [comment]
Hey @DamienGilliard I will work directly into this branch if ok for you for the python wraps I believe it's simpler than creating another one. 👍

**Comment by 9and3** [comment]
Hey [@DamienGilliard](https://github.com/DamienGilliard) I will work directly into this branch if ok for you for the python wraps I believe it's simpler than creating another one. :+1:

**Comment by 9and3** [comment]
I just merged the main in [58daa2a](https://github.com/diffCheckOrg/diffCheck/pull/23/commits/58daa2ae1266157c71a99116279c0ae105422956) (it was a bit outdated this branch for doing the wraps), I think the easiest would be to not branching  out for the wraps. If ok for you I will do it here directly and wait for your PR in #33 .

**Comment by 9and3** [comment]
Hello @DamienGilliard ,

I refactored/modified a bit the segmentation backend (mainly small fixes and decoupled some things from the segmentation function). I added two components: a) the normal estimator + b) the normal segmentation.

gh_snap5

Pop a message when the scan segmentation is ready for review on the other PR! 👍

**Comment by DamienGilliard** [comment]
Et merci !!

**Comment by 9and3** [comment]
Hello @DamienGilliard ! As discussed here's the 🥔 ..
You can find an example file with internalized data I was testing [here](https://github.com/diffCheckOrg/diffCheck/blob/segmentation/src/gh/examples/assembly_beam_system.ghx).

The metadata and wrap is there, just some work from c++ is needed.

As a note:
- the cloud is passing as reference
- with a threshold of app. 0.4 is giving the result from the image and only a very little number of clusters are associated

image

Let me know if you can make the associator a bit more tolerant to worse normal and less populated points as the one internalized the gh file.. thanks` 🦖

**Comment by DamienGilliard** [comment]
@9and3 I'm on it !

**Comment by DamienGilliard** [comment]
Capture d'écran 2024-07-16 214321

[WIP]:
- the point association has been changed and works better.
- Some unnneeded deletions on the c++ side were causing trouble and have been removed

**Comment by 9and3** [comment]
> * find a way to remove unwanted points that are now associated as a result of the new point association technique:

this point can be achieved by euclidean clustering/filtering maybe

**Comment by 9and3** [comment]
Ready for merging @DamienGilliard ?

File: src/diffCheck/geometry/DFMesh.hh
Comment (9and3) [inline src/diffCheck/geometry/DFMesh.hh:1]: maybe here we could add the function to test if a point is on a mesh face @DamienGilliard ?

File: src/diffCheck/geometry/DFMesh.cc [hunk 1]
```diff
@@ -79,6 +79,62 @@ namespace diffCheck::geometry
...
+            // arbitrary value to avoid false positives (points that, when projected on the triangle, are in it, but that are actually located too far from the mesh to actually belong to it)
+            double maxProjectionDistance = std::min({(v1 - v0).norm(), (v2 - v1).norm(), (v0 - v2).norm()});
+
+            if (std::abs(res) < associationThreshold && (projectedPoint - point).norm() < maxProjectionDistance)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
```
Comment (DamienGilliard) [inline src/diffCheck/geometry/DFMesh.cc:136]: @9and3 It's there !

**Review by DamienGilliard — COMMENTED** [review]
your suggestion was implemented ;)

**Pull request merged**